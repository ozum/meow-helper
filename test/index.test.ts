import chalk from "chalk";
import getHelp from "../src/index";
import type { ExtendedAnyFlags } from "../src/get-help";

chalk.level = 0;

const cliFlags: ExtendedAnyFlags = {
  cwd: { alias: "c", type: "string", desc: "Set current working directory for relative paths." },
  ignoreConfigs: {
    alias: "i",
    type: "string",
    default: "node_modules,dist",
    desc: "(CSV) Ignore configuration files (e.g. .gitignore, .prettierignore) to add new created files if any.",
  },
  dry: { alias: "d", type: "boolean", desc: "Prevents changes to be written to disk. Executes a dry run." },
  verbose: { alias: "v", type: "boolean", default: true, desc: "Outputs extra information." },
  linkSameDir: {
    type: "boolean",
    desc: "Move files near original one for iCloudDrive. For example 'dist' is moved 'dist.nosync' in same directory.",
  },
  colors: { alias: "c", type: "string", isMultiple: true, isRequired: true, desc: "Favorite colors." },
  size: { alias: "s", type: "number", isRequired: () => true, desc: "Size." },
};

describe("getHelp()", () => {
  it("should throw if command cannot be found.", () => {
    expect(() => getHelp({ autoHelp: false, args: { path: "some" } })).toThrow("Either 'command'");
  });

  it("should return help text.", () => {
    const help = getHelp({
      multilineThreshold: 50,
      command: "not-sync",
      lineLength: 80,
      usage: ["not-sync [options] <path>...", "not-sync <path>..."],
      description: "Disable file synchronization for files in an auto detected cloud storage such as Dropbox, iCloudDrive or OneDrive.",
      args: { "path*": "Path or list of paths to disable syncronization for.", cli: "use cli" },
      flags: cliFlags,
      examples: "not-sync node_modules,dist,coverage",
      groups: { cwd: { title: "General Options", description: "General description." }, colors: { title: "Appereance Options" } },
    });
    expect(help).toMatchSnapshot();
  });

  it("should render withot any alias.", () => {
    const help = getHelp({ command: "not-sync", flags: { color: { desc: "color" }, size: { desc: "size" } } });
    expect(help).toMatchSnapshot();
  });

  it("should use default values.", () => {
    const help = getHelp({ flags: cliFlags, command: "not-sync" });
    expect(help).toMatchSnapshot();
  });

  it("should use package.json data.", () => {
    const help = getHelp({ flags: cliFlags, pkg: { name: "not-sync", description: "details" } });
    expect(help).toMatchSnapshot();
  });

  it("should generate help without options.", () => {
    const help = getHelp({ args: { path: "some" }, command: "not-sync" });
    expect(help).toMatchSnapshot();
  });

  it("should generate name details if `autoHelp` is false.", () => {
    const help = getHelp({ autoHelp: false, command: "not-sync", description: "Desc" });
    expect(help).toMatchSnapshot();
  });

  it("should generate name details without command and description if `autoHelp` is false.", () => {
    const help = getHelp({ autoHelp: false, command: "not-sync", args: { "path...": "some" } });
    expect(help).toMatchSnapshot();
  });

  it("should throw if process exit with code 2.", () => {
    getHelp({ command: "not-sync", notThrow: false });
    expect(true).toBe(true);
  });

  it("should generate options without any default value.", () => {
    const help = getHelp({ command: "not-sync", flags: { path: { desc: "a" }, cwd: { desc: "b" } } });
    expect(help).toMatchSnapshot();
  });
});
