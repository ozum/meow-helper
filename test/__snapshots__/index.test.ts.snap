// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHelp() should generate help without options. 1`] = `
"
.------------------------------------------------------------------------------.
| [1mnot-sync[22m                                                                     |
'------------------------------------------------------------------------------'

[1m[7m[32mÂ Â Â Â Â USAGEÂ Â Â Â Â [39m[27m[22m

[2m$[22m [32mnot-sync[39m [36mpath[39m

[1m[7m[36mÂ Â Â ARGUMENTSÂ Â Â [39m[27m[22m

[36m<path>[39m some
"
`;

exports[`getHelp() should return help text. 1`] = `
"
.----------------------------------------------------------------------------------------------------------------------.
| [1mnot-sync[22m                                                                                                             |
'----------------------------------------------------------------------------------------------------------------------'
Disable file synchronization for files in an auto detected cloud storage such as Dropbox, iCloudDrive or OneDrive.

[1m[7m[32mÂ Â Â Â Â USAGEÂ Â Â Â Â [39m[27m[22m

[2m$[22m [32mnot-sync[39m [options] <path>...
[2m$[22m [32mnot-sync[39m <path>...

[1m[7m[36mÂ Â Â ARGUMENTSÂ Â Â [39m[27m[22m

[36m<path>[39m Path or list of paths to disable syncronization for.
[36m<cli>[39m  use cli

[1m[7m[33mÂ Â Â Â OPTIONSÂ Â Â Â [39m[27m[22m

[33m-c --cwd[39m                                        Set current working directory for relative paths.
[33m-i --ignoreConfigs[39m [2m(Default: [22m[33mnode_modules,dist[39m[2m)[22m (CSV) Ignore configuration files (e.g. .gitignore, .prettierignore) to
                                                add new created files if any.
[33m-d --dry[39m                                        Prevents changes to be written to disk. Executes a dry run.
[33m-v --verbose[39m       [2m(Default: [22m[33mtrue[39m[2m)[22m              Outputs extra information.
[33mÂ Â  --linkSameDir[39m                                Move files near original one for iCloudDrive. For example 'dist' is
                                                moved 'dist.nosync' in same directory.

[1m[7m[35mÂ Â Â Â EXAMPLESÂ Â Â [39m[27m[22m

[2m$[22m [32mnot-sync[39m node_modules,dist,coverage
"
`;

exports[`getHelp() should use default values. 1`] = `
"
[1m[7m[33mÂ Â Â Â OPTIONSÂ Â Â Â [39m[27m[22m

[33m-c --cwd[39m
     Set current working directory for relative paths.
[33m-i --ignoreConfigs[39m                                  [2m(Default: [22m[33mnode_modules,dist[39m[2m)[22m
     (CSV) Ignore configuration files (e.g. .gitignore, .prettierignore) to add
     new created files if any.
[33m-d --dry[39m
     Prevents changes to be written to disk. Executes a dry run.
[33m-v --verbose[39m                                                     [2m(Default: [22m[33mtrue[39m[2m)[22m
     Outputs extra information.
[33mÂ Â  --linkSameDir[39m
     Move files near original one for iCloudDrive. For example 'dist' is moved
     'dist.nosync' in same directory.
"
`;

exports[`getHelp() should use package.json data. 1`] = `
"
.------------------------------------------------------------------------------.
| [1mnot-sync[22m                                                                     |
'------------------------------------------------------------------------------'
details

[1m[7m[32mÂ Â Â Â Â USAGEÂ Â Â Â Â [39m[27m[22m

[2m$[22m [32mnot-sync[39m [33m<options>[39m

[1m[7m[33mÂ Â Â Â OPTIONSÂ Â Â Â [39m[27m[22m

[33m-c --cwd[39m
     Set current working directory for relative paths.
[33m-i --ignoreConfigs[39m                                  [2m(Default: [22m[33mnode_modules,dist[39m[2m)[22m
     (CSV) Ignore configuration files (e.g. .gitignore, .prettierignore) to add
     new created files if any.
[33m-d --dry[39m
     Prevents changes to be written to disk. Executes a dry run.
[33m-v --verbose[39m                                                     [2m(Default: [22m[33mtrue[39m[2m)[22m
     Outputs extra information.
[33mÂ Â  --linkSameDir[39m
     Move files near original one for iCloudDrive. For example 'dist' is moved
     'dist.nosync' in same directory.
"
`;
